File: 자바/chap20_jdbc/src/jdbc_roll_allocation/BookDAO.java
package jdbc_roll_allocation;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;

public class BookDAO implements IDao<Book, Integer>{

	@Override
	public int insert(Book vo) throws SQLException {
		String sql = "INSERT INTO BOOK(bookid,bookname,publicher,price) "
				+ "VALUES(?,?,?,?)";
		Connection conn = ConnectionFactory.create();	// 커넥션 항상 선언하는건 그때그때닫아주기 위해서(부하를 막아줌)
		// ? 표가 없어도 되는 질의 - 정적인 질의를 사용할 때에는 Statement
		// ? 표가 있는 질의  - 동적인 질의를 사용할 때에는 PreparedStatement 사용
		PreparedStatement pstmt = conn.prepareStatement(sql);	// 질의를 실어 보낼 객체
		
		pstmt.setInt(1,vo.getBookid());
		pstmt.setString(2, vo.getBookname());
		pstmt.setString(3, vo.getPublicher());
		pstmt.setInt(4, vo.getPrice());	// 질의를 실어 보낼 객체 완성
		//////////////////////////////////////////////
		int res = pstmt.executeUpdate();	// 질의 전송, 수행을 못한 경우 -1이 옴
		conn.close();	// 잊지말자!
		return res;
	}

	@Override
	public int delete(Integer key) throws SQLException {
		String sql = "DELETE BOOK WHERE BOOKID = ?";
		Connection conn = ConnectionFactory.create();
		PreparedStatement pstmt = conn.prepareStatement(sql);
		pstmt.setInt(1, key);
		int res = pstmt.executeUpdate();
		conn.close();
		return res;
	}

	@Override
	public int update(Book vo) throws SQLException {
		Connection conn = ConnectionFactory.create();
		String sql = "UPDATE BOOK SET BOOKNAME = ?, PUBLICHER = ?, PRICE = ? "
				+ " WHERE BOOKID = ? ";
		PreparedStatement pstmt = conn.prepareStatement(sql);
		pstmt.setString(1, vo.getBookname());
		pstmt.setString(2, vo.getPublicher());
		pstmt.setInt(3, vo.getPrice());
		pstmt.setInt(4, vo.getBookid());
		int res = pstmt.executeUpdate();
		conn.close();
		return res;
	}

	@Override	// 자료를 찾을 때 사용
	public Book select(Integer key) throws SQLException {
		Connection conn = ConnectionFactory.create();
		String sql = "SELECT * FROM BOOK WHERE BOOKID = ? ";	// * = 전부 다
		PreparedStatement pstmt = conn.prepareStatement(sql);
		pstmt.setInt(1, key);
		ResultSet rs = pstmt.executeQuery();	// 여러개의 레코드 정보가 리턴 될 수 있음
		Book vo = new Book();	// 빈껍데기 기본생성자 만듦.
		while(rs.next()) {	// NEXT = 커서 옮기는것
			vo.setBookid(rs.getInt("bookid"));	//1(위치 바뀔수 있어서 이름쓰는게 더 안전)
			vo.setBookname(rs.getNString("bookname"));
			vo.setPublicher(rs.getNString("publicher"));
			vo.setPrice(rs.getInt("price"));
		}
		conn.close();
		return vo;
	}

	@Override
	public List<Book> selectAll() throws SQLException {
		Connection conn = ConnectionFactory.create();
		String sql = "SELECT * FROM BOOK ";	// * = 전부 다
		PreparedStatement pstmt = conn.prepareStatement(sql);	// ? 없더라도 쓰면 편함
		ResultSet rs = pstmt.executeQuery();	// 여러개의 레코드 정보가 리턴 될 수 있음
		List<Book> list = new ArrayList<>();
		while(rs.next()) {
			Book vo = new Book();	// 빈껍데기 기본생성자 만듦.
			vo.setBookid(rs.getInt("bookid"));	//1(위치 바뀔수 있어서 이름쓰는게 더 안전)
			vo.setBookname(rs.getNString("bookname"));
			vo.setPublicher(rs.getNString("publicher"));
			vo.setPrice(rs.getInt("price"));
			list.add(vo);
		}
		conn.close();
		return list;
	}

	@Override	// conditions 작성 예 = "WHERE PRICE>7000"
	public List<Book> selectByConditions(String conditions) throws SQLException {
		Connection conn = ConnectionFactory.create();
		String sql = "SELECT * FROM BOOK "+conditions;	// 컨디션에 WHERE 포함해야함
		PreparedStatement pstmt = conn.prepareStatement(sql);	// ? 없더라도 쓰면 편함
		ResultSet rs = pstmt.executeQuery();	// 여러개의 레코드 정보가 리턴 될 수 있음
		List<Book> list = new ArrayList<>();
		while(rs.next()) {
			Book vo = new Book();	// 빈껍데기 기본생성자 만듦.
			vo.setBookid(rs.getInt("bookid"));	//1(위치 바뀔수 있어서 이름쓰는게 더 안전)
			vo.setBookname(rs.getNString("bookname"));
			vo.setPublicher(rs.getNString("publicher"));
			vo.setPrice(rs.getInt("price"));
			list.add(vo);
		}
		conn.close();
		return list;
	}

	@Override
	public int getMaxNum() throws SQLException {
		Connection conn = ConnectionFactory.create();
		String sql = "SELECT MAX(BOOKID) FROM BOOK ";
		PreparedStatement pstmt = conn.prepareStatement(sql);
		ResultSet rs = pstmt.executeQuery();
		int max = 0;
		while (rs.next()) {
			max = rs.getInt(1);
		}
		conn.close();
		return max;
	}

	@Override
	public int getMinNum() throws SQLException {
		Connection conn = ConnectionFactory.create();
		String sql = "SELECT MIN(BOOKID) FROM BOOK ";
		PreparedStatement pstmt = conn.prepareStatement(sql);
		ResultSet rs = pstmt.executeQuery();
		int min = 0;
		while (rs.next()) {
			min = rs.getInt(1);
		}
		conn.close();
		return min;
	}
}

File: 파이썬/Python_AI_wxpython/src/oracle_ex/BookDAO.py
'''
Created on 2021. 7. 23.

@author: pc354
'''
import time

import cx_Oracle


class BookDAO(): #BOOKID, BOOKNAME, PUBLISHER, PRICE

    def __init__(self):
        pass

    def insert(self,vo,conn):
        cur = conn.cursor()
        sql = '''
            INSERT INTO BOOK VALUES(:1,:2,:3,:4)
        '''
        cur.execute(sql,vo)
        conn.commit()
        cur.close()
        conn.close()
        print('1건의 자료 입력 완료')
        
    def select(self,key,conn):
        cur = conn.cursor()
        sql = '''
            SELECT * FROM BOOK WHERE BOOKID = :1 
        '''
        cur.execute(sql,key)
        data = cur.fetchone()
        conn.commit()
        cur.close()
        conn.close()
        print('1건의 자료 색인 완료')
        return data
    
    def selectAll(self,conn):
        cur = conn.cursor()
        sql = '''
            SELECT * FROM BOOK
            '''
        cur.execute(sql)
        data = cur.fetchall()
        conn.commit()
        cur.close()
        conn.close()
        return data
    
    def delete(self,key,conn):
        cur = conn.cursor()
        sql = '''
            DELETE FROM BOOK WHERE BOOKID = :1 
            '''
        cur.execute(sql,key)
        conn.commit()
        cur.close()
        conn.close()
        return data
    
    # def update(self,vo,conn):
    #     cur = conn.cursor()
    #     sql = '''
    #         UPDATE BOOK SET BOOKNAME = :2, PUBLISHER = :3, PRICE = :4 WHERE BOOKID = :1
    #     '''
    #     cur.execute(sql,vo)
    #     conn.commit()
    #     cur.close()
    #     conn.close()
    #     print('자료 수정 완료')
        
    def update(self,vo,conn):
        cur = conn.cursor() 
        sql = '''
             UPDATE BOOK SET BOOKNAME = '%s', PUBLISHER = '%s' , PRICE =%d 
             WHERE BOOKID = %d
        '''%(vo[1],vo[2],vo[3],vo[0])
        print(vo)
        cur.execute(sql)
        conn.commit()
        
        cur.close() 
        conn.close()
        print("자료 수정 완료")   
    
if __name__ == '__main__':
    obj = BookDAO()
    # conn = cx_Oracle.connect('madang/madang@localhost:1521/xe')
    # obj.insert((16,'돌아와요부산항','용필출판',80000),conn)
    
    # conn = cx_Oracle.connect('madang/madang@localhost:1521/xe')
    # data = obj.select((16,),conn)
    # print(data) 
    
    # conn = cx_Oracle.connect('madang/madang@localhost:1521/xe')
    # data = obj.delete((16,), conn)
    # print('데이터 삭제 완료~!')
          
    # vo = (1,'축구의 역사','굿스포츠',7000)
    # conn = cx_Oracle.connect('madang/madang@localhost:1521/xe')
    # obj.update(vo, conn)
    # time.sleep(0.5)

    
    
    conn = cx_Oracle.connect('madang/madang@localhost:1521/xe')
    data = obj.selectAll(conn)
    for row in data:
        print(row)